Pw/pip windows bug

pwelch opened the pull request:
When pip on windows has a newer version available it spits out a stderr message along with the results of the `pip show` command. 

This adds logic to ignore the pip message that there is a newer version. 

Fixes #2855

Commit:
Add python check for pip resource

When checking pip resources, we should skip resource if python is not
installed or we will fail with an error when trying to parse the path.

Commit:
Check pip command on windows

On Windows, if pip has a newer version available, it adds an error
message to stderr. Now checking if both stderr and stdout on windows
have values. If so, assume pip package is installed.

Fixes #2855

Signed-off-by: Paul Welch <pwelch@chef.io>

jquick commented on lib/resources/pip.rb:
```diff
@@ -62,18 +63,52 @@ def to_s
+    end
+
+    def cmd_successful?
+      result = inspec.command("#{@pip_cmd} show #{@package_name}")
```
Could we just use `cmd` here instead of doing the `inspec.command` again?

pwelch replied:
Fixed

jquick commented on lib/resources/pip.rb:
```diff
@@ -62,18 +63,52 @@ def to_s
+    # {"Pip" => nil, "Python" => "/path/to/python"}
+    #
+    # @return [Hash] of paths
+    def paths
```
You mind updating this to `windows_paths`. Since its windows only command it may read better.

pwelch replied:
@jquick good catch! Thanks!

jquick reviewed (commented):
Really nice work here @pwelch ! Just a few comments.

Commit:
Refactor recent pip changes for Windows

Signed-off-by: Paul Welch <pwelch@chef.io>

TrevorBramble commented on lib/resources/pip.rb:
```diff
@@ -26,16 +26,16 @@ class PipPackage < Inspec.resource(1)
       @info[:type] = 'pip'
-      cmd = inspec.command("#{@pip_cmd} show #{@package_name}")
-      return @info if cmd.exit_status != 0
+      return @info if !cmd_successful?
```
```ruby
return @info unless cmd_successful?
```

TrevorBramble commented on lib/resources/pip.rb:
```diff
@@ -62,19 +62,56 @@ def to_s
+    end
+
+    def cmd_successful?
+      result = cmd
```
Seems like every instance of `result` in this method could just be `cmd`? (And would actually read better that way?)

TrevorBramble commented on lib/resources/pip.rb:
```diff
@@ -62,19 +62,56 @@ def to_s
+    # @return [Hash] of windows_paths
+    def windows_paths
+      return @__windows_paths if @__windows_paths
+      cmd = inspec.command('New-Object -Type PSObject | Add-Member -MemberType NoteProperty -Name Pip -Value (Invoke-Command -ScriptBlock {where.exe pip}) -PassThru | Add-Member -MemberType NoteProperty -Name Python -Value (Invoke-Command -ScriptBlock {where.exe python}) -PassThru | ConvertTo-Json')
```
Can we break this out onto multiple lines?
```ruby
cmd = inspec.command(
  'New-Object -Type PSObject' \
  ' | Add-Member -MemberType NoteProperty -Name Pip -Value (Invoke-Command -ScriptBlock {where.exe pip}) -PassThru' \
  ' | Add-Member -MemberType NoteProperty -Name Python -Value (Invoke-Command -ScriptBlock {where.exe python}) -PassThru' \
  ' | ConvertTo-Json')
```
or better...
```ruby
cmd = inspec.command([
  'New-Object -Type PSObject',
  'Add-Member -MemberType NoteProperty -Name Pip -Value (Invoke-Command -ScriptBlock {where.exe pip}) -PassThru',
  'Add-Member -MemberType NoteProperty -Name Python -Value (Invoke-Command -ScriptBlock {where.exe python}) -PassThru',
  'ConvertTo-Json'
].join(' | '))

TrevorBramble reviewed (commented):
Some nitpicks

Commit:
Clean up powershell query command

- Make it easier to read what the powershell command is doing
- Make it easier to read what the cmd_successful method lokos for

Signed-off-by: Paul Welch <pwelch@chef.io>

jquick reviewed (approved):
Thanks @pwelch !
